Port to gophercloud v2

vooon opened the pull request:
That is a refactoring to use `gophercloud/v2`.

- Update Go requirement to 1.25
- Drop gophercloud v1
- Fix #494 
- Added API discovery reports fixture to most of the services
- Add test for Swift
- Add build_info metric

Tries to solve some of #500

vooon committed:
port(nova): migrate to gophercloud v2

Why: mechanical service client migration.

Signed-off-by: Vladimir Ermakov <vooon341@gmail.com>

vooon committed:
port(cinder): migrate to gophercloud v2

Why: mechanical service client migration.

Signed-off-by: Vladimir Ermakov <vooon341@gmail.com>

vooon committed:
port(glance): migrate to gophercloud v2

Why: mechanical service client migration.

Signed-off-by: Vladimir Ermakov <vooon341@gmail.com>

vooon committed:
port(keystone): migrate to gophercloud v2

Why: mechanical service client migration.

Signed-off-by: Vladimir Ermakov <vooon341@gmail.com>

vooon committed:
port(framework): standardize list function signatures

Why: mechanical signature alignment for gophercloud v2.

Signed-off-by: Vladimir Ermakov <vooon341@gmail.com>

vooon committed:
port(neutron): migrate to gophercloud v2

Why: mechanical service client migration.

Signed-off-by: Vladimir Ermakov <vooon341@gmail.com>

Sharpz7 commented on exporters/nova.go:
```diff
@@ -117,6 +107,13 @@ var defaultNovaMetrics = []Metric{
 func NewNovaExporter(config *ExporterConfig, logger *slog.Logger) (*NovaExporter, error) {
+	ctx := context.TODO()
+
+	err := utils.SetupClientMicroversionV2(ctx, config.ClientV2, "OS_COMPUTE_API_VERSION", novaLatestSupportedMicroversion)
```
It could also be that the fixtures need updated https://github.com/openstack-exporter/openstack-exporter/blob/main/exporters/exporter_test.go#L56

vooon commented:
@Sharpz7 sorry, i was on a vacation. I'm unsure if i can fully test things like Heat or Designate, as the stacks i'm currently working on have quite a narrow use.
Of course i can setup devstack, just it'll be far from what i had working on full featured distro (at the same time, i'm glad not:).

Sharpz7 commented:
@vooon I don't have that either, but the CI is pretty good right now. If you don't mind, I can work on this PR after I get the integration tests working (that sets up a devstack in CI so that we get better tests).

Just let me know if you are okay with me pushing to this branch

Sharpz7 commented:
Sounds good - might take me a while to get to but hopefully not too long, I think you have done a large chunk of the work already :)

vooon committed:
support OS_DEBUG to log gophercloud queries

Signed-off-by: Vladimir Ermakov <vooon341@gmail.com>

vooon commented:
@Sharpz7 some updates:

- Updated nova hypervisors with the one I got from the devstack (but then i editied values to be the same).
  Mainly that was required because of incorrect next page link.
  It was a bit unpleasant to find out why the heck gophercloud tries to pull some unrelated url.
 
- Added request logging is `OS_DEBUG` env set (what helped me to debug trouble above).
- Fixed microversion requirement for Nova. Unsure why that worked in fleeting runner... But now, when API responds with multiple versions, and you don't have `/v2.1` suffix - it breaks.
  That's actually not the case for recent clouds, where you usually has only `/v2.1`...
  
- Updated gophercloud to 2.9.0 and fixed vCPU counting bug. But that enforced to change Go to 1.24.

vooon committed:
port(placement): migrate to gophercloud v2

Why: mechanical service client migration.

Signed-off-by: Vladimir Ermakov <vooon341@gmail.com>

vooon committed:
port(object-store): migrate to gophercloud v2

Why: mechanical service client migration.

Signed-off-by: Vladimir Ermakov <vooon341@gmail.com>

vooon committed:
port(designate): migrate to gophercloud v2

Why: mechanical service client migration.

Signed-off-by: Vladimir Ermakov <vooon341@gmail.com>

vooon committed:
port(octavia): migrate to gophercloud v2

Why: mechanical service client migration.

Signed-off-by: Vladimir Ermakov <vooon341@gmail.com>

vooon committed:
port(manila): migrate to gophercloud v2

Why: mechanical service client migration.

Signed-off-by: Vladimir Ermakov <vooon341@gmail.com>

vooon committed:
test: consolidate fixtures and service tests

Why: keep fixture/test churn in one place while port commits stay mechanical.

Signed-off-by: Vladimir Ermakov <vooon341@gmail.com>

vooon committed:
port(gnocchi): migrate to gophercloud v2

Why: mechanical service client migration.

Signed-off-by: Vladimir Ermakov <vooon341@gmail.com>

vooon committed:
port(cache): migrate cache path to gophercloud v2

Why: mechanical migration to gophercloud v2 APIs.

Signed-off-by: Vladimir Ermakov <vooon341@gmail.com>

Sharpz7 commented:
Yes @vooon! I view this AI stuff as "the worst it will ever be" - I think eventually it will be really helpful!

Meanwhile I'll work on #481

vooon committed:
nova: add tenant_id to quota labels

Why: prevent project-name collisions (cardinality).

Signed-off-by: Vladimir Ermakov <vooon341@gmail.com>

vooon committed:
neutron: add tenant_id to quota labels

Why: prevent project-name collisions (cardinality).

Signed-off-by: Vladimir Ermakov <vooon341@gmail.com>

vooon committed:
main: refactor cancellation, expose program version

Signed-off-by: Vladimir Ermakov <vooon341@gmail.com>

vooon committed:
neutron: add device_id on port metric

Why: alert correlation improvement (cardinality).

Signed-off-by: Vladimir Ermakov <vooon341@gmail.com>

vooon committed:
fix(nova): handle hypervisors pagination >1000

Fix #519 which is proposed in #520

Co-authored-by: Dejan Sanader <dejan.sanader@ovhcloud.com>
Signed-off-by: Vladimir Ermakov <vooon341@gmail.com>

Sharpz7 commented:
https://github.com/openstack-exporter/openstack-exporter/pull/520 from @dsanader is here

vooon committed:
go.mod: require go 1.26

Signed-off-by: Vladimir Ermakov <vooon341@gmail.com>

vooon committed:
ironic: expose maintenance_reason label

Why: improve troubleshooting context for baremetal.

Signed-off-by: Vladimir Ermakov <vooon341@gmail.com>

vooon committed:
docs: sync README metrics and CLI flags

Summary:
- align README metrics table and CLI flags with current behavior
- add exporter build info metric entry

Why:

Keep documentation in sync with implementation to avoid operator confusion.

Signed-off-by: Vladimir Ermakov <vooon341@gmail.com>

vooon committed:
ironic: expose update/provision timestamps

Why: add operational visibility for node lifecycle.

Signed-off-by: Vladimir Ermakov <vooon341@gmail.com>

vooon committed:
placement: add resource generation metric

Why: expose provider generation for drift diagnostics.

Signed-off-by: Vladimir Ermakov <vooon341@gmail.com>

vooon committed:
auto-detect enabled services in single-cloud

Why: select available services and fail fast on auth/catalog issues.

Signed-off-by: Vladimir Ermakov <vooon341@gmail.com>

vooon committed:
port(exporter): remove remaining gophercloud v1 paths

Why: mechanical framework migration to gophercloud v2.

Signed-off-by: Vladimir Ermakov <vooon341@gmail.com>

vooon committed:
fix(cache): fix content type header

Signed-off-by: Vladimir Ermakov <vooon341@gmail.com>

vooon committed:
port(orchestration): migrate to gophercloud v2

Why: mechanical service client migration.

Signed-off-by: Vladimir Ermakov <vooon341@gmail.com>

vooon committed:
port(trove): migrate to gophercloud v2

Why: mechanical service client migration.

Signed-off-by: Vladimir Ermakov <vooon341@gmail.com>

vooon committed:
port(magnum): migrate to gophercloud v2

Why: mechanical service client migration.

Signed-off-by: Vladimir Ermakov <vooon341@gmail.com>

vooon committed:
port(framework): add shared microversion setup helper

Why: centralize microversion setup during gophercloud v2 porting.

Signed-off-by: Vladimir Ermakov <vooon341@gmail.com>

vooon committed:
port(ironic): use shared microversion setup helper

Why: move ironic microversion setup to the common framework helper.

Signed-off-by: Vladimir Ermakov <vooon341@gmail.com>

Sharpz7 commented on utils/microversion.go:
```diff
@@ -0,0 +1,84 @@
+package utils
```
https://github.com/openstack/kolla-ansible/commit/41fba3c5df8880a899ddee2e47218c0eab1778b7

Looking at this, should we not already support microversions? I do not understand this 100% (how it seems to work already since v1.5)

Sharpz7 committed:
Merge https://github.com/openstack-exporter/openstack-exporter into port-to-gc-v2

Sharpz7 commented on exporters/cinder.go:
```diff
@@ -280,30 +244,19 @@ func ListCinderPoolCapacityFree(exporter *BaseOpenStackExporter, ch chan<- prome
 	}
 
-	allPagesProject, err := projects.List(c, projects.ListOpts{DomainID: exporter.DomainID}).AllPages()
+	allPagesProject, err := projects.List(cli, projects.ListOpts{DomainID: exporter.DomainID}).AllPages(ctx)
```
Is this code repeating things from https://github.com/openstack-exporter/openstack-exporter/pull/424?

vooon replied:
Not really, the only common thing - that keystone client creation moved to a common func.
But not carry anything related to tenant_id.

Sharpz7 commented on exporters/ironic.go:
```diff
@@ -5,22 +5,38 @@ import (
 func NewIronicExporter(config *ExporterConfig, logger *slog.Logger) (*IronicExporter, error) {
+	ctx := context.TODO()
+
+	// NOTE(Sharpz7) Gophercloud V2 adds this new field ResourceBase.
```
Did you see this issue yourself / in other services? cc @jayofdoom

vooon replied:
No, it's working fine without ResourceBase manipulation with Kolla.

Sharpz7 commented:
In general, I am really happy with this. Waiting on https://github.com/openstack-exporter/openstack-exporter/pull/481 being merged, make sure they pass against this. 

Thank you for your work!

Sharpz7 committed:
Merge branch 'main' into port-to-gc-v2

Sharpz7 committed:
Attempting to fix compute integration tests by making them more stable

Sharpz7 committed:
Test improvements

Sharpz7 committed:
Fixed parser

Sharpz7 committed:
Merge branch 'main' into port-to-gc-v2

cardoe commented on .github/workflows/ci.yaml:
```diff
@@ -85,7 +85,7 @@ jobs:
         uses: actions/setup-go@v6
         with:
-          go-version: 1.23
+          go-version: 1.26
```
As a follow on, put the version in the go.mod and just let the action read it from there.

cardoe commented on Dockerfile:
```diff
@@ -1,4 +1,4 @@
-FROM golang:1.23 AS build
+FROM golang:1.26 AS build
```
We can parametrize this in a follow on and pass it as part of the build from go.mod.

cardoe commented on README.md:
```diff
@@ -118,41 +118,49 @@ Flags:
       --[no-]cache               Enable Cache mechanism globally
       --cache-ttl=300s           TTL duration for cache expiry(eg. 10s, 11m, 1h)
-
+      --tenant-id=TENANT-ID      Gather metrics only for the given Tenant ID
```
We should switch to project-id in a follow on.

Sharpz7 committed:
Merge remote-tracking branch 'upstream/main' into port-to-gc-v2

# Conflicts:
#	cache/utils.go
#	exporters/cinder.go
#	exporters/designate.go
#	exporters/keystone.go
#	exporters/neutron.go
#	exporters/nova.go
#	exporters/utils.go
#	go.mod
#	main.go

Sharpz7 committed:
Merge remote-tracking branch 'voon/port-to-gc-v2' into port-to-gc-v2

cardoe reviewed (approved):
Addressed my concerns. Just need to get tests going.

vooon commented:
I'd prefer if we merge that MR and continue with further improvements in a separate smaller ones.
As it really begin to be unreviewable...

vooon committed:
fix(ci): use go version from go.mod

Signed-off-by: Vladimir Ermakov <vooon341@gmail.com>

vooon committed:
dep(go.mod): update gophercloud to v2.12.0

Signed-off-by: Vladimir Ermakov <vooon341@gmail.com>

vooon committed:
placement: add resource_provider_allocations metric

Add a new gauge metric that exposes per-consumer resource allocations
from each resource provider. This enables detecting inconsistencies
between Placement DB and Nova by joining with nova_server_status
on consumer_id == server UUID.

The metric is collected in the same resource providers loop to avoid
duplicate API calls. It is guarded by a metric-existence check so
--disable-metric placement-resource_provider_allocations skips the
allocations API call entirely.

Signed-off-by: Vladimir Ermakov <vooon341@gmail.com>

vooon committed:
placement: rename consumer_id label to uuid

Matches the label name used by nova and libvirt exporters, making
PromQL joins straightforward without label_replace.

Signed-off-by: Vladimir Ermakov <vooon341@gmail.com>

Sharpz7 commented:
I agree @voon it needs to get merged now.

Sharpz7 commented:
This is a huge commit, I want to give @mnaser and @niedbalski a chance to review before it merges.

This PR however has been up for a very long time, and they have had plenty time to review before now. Therefore, I am going to say that at EOD Friday May 29th (6pm EST) I will merge it. 

Nothing else will be merged until then. @vooon thank you for this substantial effort!

cardoe commented:
Yeah this needs to merge cause there's no reviewing this.
